// state-utils/webpack.config.test.js
// See https://github.com/petehunt/webpack-howto to learn about
// these settings

var path = require('path');
var webpack = require('webpack');

var commonsPlugin =
  new webpack.optimize.CommonsChunkPlugin('common.js');

module.exports = {
  entry: {

    'test-buckets': './src/test/test-buckets',
    'test-callbacks': './src/test/test-callbacks',
    'test-csv': './src/test/test-csv',
    'test-events': './src/test/test-events',
    'test-hash-change': './src/test/test-hash-change',
    'test-loading': './src/test/test-loading',
    'test-messaging': './src/test/test-messaging',
    'test-net': './src/test/test-net',
    'test-promise': './src/test/test-promise',
    'test-react-lifecycle': './src/test/test-react-lifecycle',
    'test-react-lists-in-state': './src/test/test-react-lists-in-state',
    'test-state': './src/test/test-state',
    'test-throttle': './src/test/test-throttle'

  },

  output: {
    filename: '[name].js',
    path: path.join(__dirname, 'build'),
    publicPath: 'build/'
  },

  module: {
    loaders: [
      { test: /\.css$/, loader: 'style!css' },
      { test: /\.less$/, loader: 'style!raw!less' },
      { test: /\.woff(2)?(\?v=[0-9]\.[0-9]\.[0-9])?$/, loader: 'url-loader?limit=10000&minetype=application/font-woff' },
      { test: /\.(ttf|eot|svg)(\?v=[0-9]\.[0-9]\.[0-9])?$/, loader: 'file-loader' },
      { test: /\.jsx$/, loaders: ['react-hot-loader', 'babel'], exclude: /node_modules/ },
      { test: /\.js$/, loaders: ['react-hot-loader', 'babel'], exclude: /node_modules/ },
      { test: /\.png$/, loaders: ['url-loader?limit=10000']},
      { test: /\.json$/, loaders: ['json'], exclude: /node_modules/ }
    ]
  },

  resolve: {
    // you can now require('file') instead of require('file.coffee')
    extensions: ['', '.js', '.jsx', '.json'] 
  },

  plugins: [commonsPlugin]
};
