// state-utils/webpack.config.js
// See https://github.com/petehunt/webpack-howto to learn about
// these settings

var path = require('path');
var webpack = require('webpack');


module.exports = {
  entry: {
    'state-utils': './src/state-utils.js',
  },

  output: {
    filename: '[name].js',
    //library: 'StateUtils',
    libraryTarget: 'commonjs2',
    path: path.join(__dirname, 'dist'),
    publicPath: 'dist/'
  },

  module: {
    loaders: [
      { test: /\.css$/, loader: 'style!css' },
      { test: /\.less$/, loader: 'style!raw!less' },
      { test: /\.woff(2)?(\?v=[0-9]\.[0-9]\.[0-9])?$/, loader: 'url-loader?limit=10000&minetype=application/font-woff' },
      { test: /\.(ttf|eot|svg)(\?v=[0-9]\.[0-9]\.[0-9])?$/, loader: 'file-loader' },
      { test: /\.jsx$/, loaders: ['babel'], exclude: /node_modules/ },
      { test: /\.js$/, loaders: ['babel'], exclude: /node_modules/ },
      { test: /\.png$/, loaders: ['url-loader?limit=10000']},
      { test: /\.json$/, loaders: ['json'], exclude: /node_modules/ }
    ]
  },

  resolve: {
    // you can now require('file') instead of require('file.coffee')
    extensions: ['', '.js', '.jsx', '.json'] 
  }
};
